agent/consul/acl_authmethod.go

```diff
@@ -136,9 +136,16 @@ func (s *Server) evaluateRoleBindings(
 			})
 
 		case structs.BindingRuleBindTypeRole:
-			roleLinks = append(roleLinks, structs.ACLTokenRoleLink{
-				Name: bindName,
-			})
+			_, role, err := s.fsm.State().ACLRoleGetByName(nil, bindName)
```

Comment:
Will the state store also reject it during the raftApply if the role doesn't exist?

In many scenarios we do a bunch of pre-raft checks but they are a little racey as by the time we call raftApply and when the log gets applied to the FSM things could have changed.

Its a lot nicer to not apply a log when we know it will fail but we still need the same protections in the state store.

Reply:
The intent is that during a Login we do a point-in-time evaluation of the binding rules to fabricate a token. In this case I'd described binding rules of `BindType="role"` as only being applicable if a role at login-time exists with that name.

Prior to this patch the state store was rejecting the token creation (rightly so) because the Token had a RoleLink to a role that did not exist. Which is fine, except the intent is that the Token still gets created, but just without the linked role.

Reply:
I guess alternatively I could thread something through the raft operation to have it disregard dead links silently, then there's no race.